# Healthcare Analytics Lab: OLTP to Star Schema

**Objective**: Analyze the OLTP schema, identify performance issues, then design and build an optimized star schema.

---

## Setup & Database Connection

In [1]:
!pip install ipython-sql sqlalchemy mysql-connector-python pymysql

  Using cached ipython_sql-0.5.0-py3-none-any.whl.metadata (17 kB)
  Using cached sqlalchemy-2.0.45-cp312-cp312-win_amd64.whl.metadata (9.8 kB)
  Using cached mysql_connector_python-9.5.0-cp312-cp312-win_amd64.whl.metadata (7.7 kB)
  Using cached prettytable-3.17.0-py3-none-any.whl.metadata (34 kB)
  Using cached ipython_genutils-0.2.0-py2.py3-none-any.whl.metadata (755 bytes)
  Using cached greenlet-3.3.0-cp312-cp312-win_amd64.whl.metadata (4.2 kB)
  Using cached typing_extensions-4.15.0-py3-none-any.whl.metadata (3.3 kB)
Using cached ipython_sql-0.5.0-py3-none-any.whl (20 kB)
Using cached sqlalchemy-2.0.45-cp312-cp312-win_amd64.whl (2.1 MB)
Using cached mysql_connector_python-9.5.0-cp312-cp312-win_amd64.whl (16.5 MB)
Using cached greenlet-3.3.0-cp312-cp312-win_amd64.whl (301 kB)
Using cached typing_extensions-4.15.0-py3-none-any.whl (44 kB)
Using cached ipython_genutils-0.2.0-py2.py3-none-any.whl (26 kB)
Using cached prettytable-3.17.0-py3-none-any.whl (34 kB)



[notice] A new release of pip is available: 25.0.1 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
%load_ext sql
# %sql sqlite:///:memory:


%sql mysql+pymysql://root:password@localhost:3306/

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [7]:
%config SqlMagic.style = '_DEPRECATED_DEFAULT'

---

## Part 1: Explore the OLTP Schema

Verify all tables exist and check their data.

In [11]:
%%sql

USE `healthcare analytics lab`;
SHOW TABLES;

 * mysql+pymysql://root:***@localhost:3306/
0 rows affected.
10 rows affected.


[('billing',),
 ('departments',),
 ('diagnoses',),
 ('encounter_diagnoses',),
 ('encounter_procedures',),
 ('encounters',),
 ('patients',),
 ('procedures',),
 ('providers',),
 ('specialties',)]

---

## Part 2: Performance Analysis - 4 Business Questions

We'll write queries to answer the 4 business questions and measure their performance.

### Question 1: Monthly Encounters by Specialty

**Goal**: For each month and specialty, show total encounters and unique patients by encounter type.

In [45]:
%%sql

EXPLAIN ANALYZE
SELECT 
    DATE_FORMAT(e.encounter_date, '%Y-%m') AS encounter_month,
    s.specialty_name,
    e.encounter_type,
    COUNT(DISTINCT e.encounter_id) AS total_encounters,
    COUNT(DISTINCT e.patient_id) AS unique_patients
FROM encounters e
INNER JOIN providers p ON e.provider_id = p.provider_id
INNER JOIN specialties s ON p.specialty_id = s.specialty_id
GROUP BY 
    DATE_FORMAT(e.encounter_date, '%Y-%m'),
    s.specialty_name,
    e.encounter_type
ORDER BY 
    encounter_month,
    s.specialty_name,
    e.encounter_type;

 * mysql+pymysql://root:***@localhost:3306/
1 rows affected.


[('-> Group aggregate: count(distinct encounters.encounter_id), count(distinct encounters.patient_id)  (actual time=43.8..51.8 rows=720 loops=1)\n    -> ... (689 characters truncated) ...           -> Index lookup on e using provider_id (provider_id=p.provider_id)  (cost=2.45 rows=9.78) (actual time=0.0199..0.0228 rows=10 loops=1000)\n',)]

**Analysis:**

- **Tables joined**: 3 (encounters → providers → specialties)
- **Number of joins**: 2

**Performance:**

- **Execution time**: 0.0518 seconds (51.8 milliseconds)
- **Estimated rows scanned**: 9,784 rows (estimated)
- **Actual rows processed**: 10,000 rows from encounters table

**Bottleneck Identified:**

- **Full sort operation**: 10,000 rows sorted before grouping (encounters × providers × specialties)
- **Nested loop joins**: 1,000 provider-specialty combinations, each triggering 10 encounter lookups (10,000 total index lookups)
- **Date calculation at query time**: `DATE_FORMAT()` computed 10,000 times during GROUP BY
- **Multiple DISTINCT counts**: Two separate distinct aggregations (encounter_id and patient_id) computed for each group
- **No pre-aggregated data**: All counts and groupings calculated at query time
- **Large intermediate result set**: Join produces 10,000 rows before aggregation reduces to 720 final groups

### Question 2: Top Diagnosis-Procedure Pairs

**Goal**: What are the most common diagnosis-procedure combinations?

In [46]:
%%sql

EXPLAIN ANALYZE
SELECT 
    d.icd10_code,
    d.icd10_description,
    pr.cpt_code,
    pr.cpt_description,
    COUNT(DISTINCT ed.encounter_id) AS encounter_count
FROM encounter_diagnoses ed
INNER JOIN diagnoses d ON ed.diagnosis_id = d.diagnosis_id
INNER JOIN encounter_procedures ep ON ed.encounter_id = ep.encounter_id
INNER JOIN procedures pr ON ep.procedure_id = pr.procedure_id
GROUP BY 
    d.icd10_code,
    d.icd10_description,
    pr.cpt_code,
    pr.cpt_description
ORDER BY 
    encounter_count DESC;

 * mysql+pymysql://root:***@localhost:3306/
1 rows affected.


[('-> Sort: encounter_count DESC  (actual time=104..104 rows=592 loops=1)\n    -> Stream results  (actual time=97.7..103 rows=592 loops=1)\n        -> G ... (1481 characters truncated) ...  -> Single-row index lookup on pr using PRIMARY (procedure_id=ep.procedure_id)  (cost=0.25 rows=1) (actual time=0.0011..0.00112 rows=1 loops=10000)\n',)]

**Analysis:**

- **Tables joined**: 4 (encounter_diagnoses → diagnoses → encounter_procedures → procedures)
- **Number of joins**: 4

**Performance:**

- **Execution time**: 0.104 seconds (104 milliseconds)
- **Estimated rows scanned**: 9,744 rows (estimated)
- **Actual rows processed**: 10,000 rows from diagnoses table

**Bottleneck Identified:**

- **Full table scan on diagnoses**: 10,000 rows scanned from diagnoses table
- **Cartesian product explosion**: Junction table joins create 10,000 intermediate rows (diagnoses × encounter_diagnoses × encounter_procedures × procedures)
- **Multiple nested loops**: 4 levels of nested loop joins, with 10,000 iterations each (40,000 total index lookups)
- **Double sort operation**: Data sorted before grouping (10,000 rows) and after aggregation (592 rows)
- **Complex GROUP BY**: Grouping by 4 text columns (2 codes + 2 descriptions)
- **DISTINCT count overhead**: Computing distinct encounter_id across 10,000 rows to deduplicate cartesian product
- **No indexed access**: Starting with full table scan instead of using indexes on junction tables

### Question 3: 30-Day Readmission Rate

**Goal**: Which specialty has the highest readmission rate?

In [47]:
%%sql

EXPLAIN ANALYZE
WITH inpatient_discharges AS (
    SELECT 
        e1.patient_id,
        e1.encounter_id AS initial_encounter_id,
        e1.discharge_date,
        e1.provider_id,
        p.specialty_id
    FROM encounters e1
    INNER JOIN providers p ON e1.provider_id = p.provider_id
    WHERE e1.encounter_type = 'Inpatient' 
      AND e1.discharge_date IS NOT NULL
),
readmissions AS (
    SELECT 
        id.patient_id,
        id.initial_encounter_id,
        id.discharge_date,
        id.specialty_id,
        e2.encounter_id AS readmission_encounter_id,
        e2.encounter_date AS readmission_date,
        DATEDIFF(e2.encounter_date, id.discharge_date) AS days_to_readmit
    FROM inpatient_discharges id
    INNER JOIN encounters e2 
        ON id.patient_id = e2.patient_id
        AND e2.encounter_type = 'Inpatient'
        AND e2.encounter_date > id.discharge_date
        AND DATEDIFF(e2.encounter_date, id.discharge_date) <= 30
)
SELECT 
    s.specialty_name,
    COUNT(DISTINCT id.initial_encounter_id) AS total_discharges,
    COUNT(DISTINCT r.readmission_encounter_id) AS readmissions_within_30days,
    ROUND(COUNT(DISTINCT r.readmission_encounter_id) * 100.0 / 
          COUNT(DISTINCT id.initial_encounter_id), 2) AS readmission_rate_percent
FROM inpatient_discharges id
LEFT JOIN readmissions r 
    ON id.initial_encounter_id = r.initial_encounter_id
INNER JOIN specialties s ON id.specialty_id = s.specialty_id
GROUP BY s.specialty_name
ORDER BY readmission_rate_percent DESC;

 * mysql+pymysql://root:***@localhost:3306/
1 rows affected.


[("-> Sort: readmission_rate_percent DESC  (actual time=45.3..45.3 rows=20 loops=1)\n    -> Stream results  (actual time=44..45.3 rows=20 loops=1)\n     ... (2715 characters truncated) ...              -> Index lookup on e2 using patient_id (patient_id=e1.patient_id)  (cost=0.25 rows=1) (actual time=0.00286..0.00343 rows=1 loops=3333)\n",)]

**Schema Analysis:**

- **Tables joined**: 3 (encounters self-joined + providers + specialties)
- **Number of joins**: Self-join on encounters + 2 additional joins

**Performance:**

- **Execution time**: 0.0453 seconds (45.3 milliseconds)
- **Estimated rows scanned**: 9,784 rows from encounters table (full table scan)
- **Actual rows processed**: 10,000 rows scanned, filtered to 3,333 inpatient discharges

**Bottleneck Identified:**

- **Full table scan on encounters**: 10,000 rows scanned, then filtered to 3,333 inpatient encounters
- **Self-join creates nested loops**: Each of 3,333 initial discharges triggers lookup for readmissions (3,333 patient_id index lookups)
- **Complex join conditions in self-join**: Multiple filters applied per iteration (encounter type, date comparison, 30-day window using `DATEDIFF()`)
- **Date calculation overhead**: `to_days()` function called twice for each potential readmission pair
- **Multiple DISTINCT counts**: Four distinct aggregations computed (2 for initial encounters, 2 for readmissions)
- **Sort before and after aggregation**: Data sorted by specialty name (3,333 rows), then by readmission rate (20 final rows)
- **Low readmission rate multiplies work**: Only 264 actual readmissions found (3,333 × 0.0792), but all encounters must be checked
- **No pre-computed readmission flags**: Must calculate readmission status for every encounter at query time

### Question 4: Revenue by Specialty & Month

**Goal**: Total allowed amounts by specialty and month.

In [43]:
%%sql

EXPLAIN ANALYZE
SELECT 
    DATE_FORMAT(e.encounter_date, '%Y-%m') AS revenue_month,
    s.specialty_name,
    COUNT(DISTINCT e.encounter_id) AS total_encounters,
    SUM(b.allowed_amount) AS total_revenue,
    ROUND(AVG(b.allowed_amount), 2) AS avg_revenue_per_encounter
FROM billing b
INNER JOIN encounters e ON b.encounter_id = e.encounter_id
INNER JOIN providers p ON e.provider_id = p.provider_id
INNER JOIN specialties s ON p.specialty_id = s.specialty_id
WHERE b.claim_status = 'Paid'
GROUP BY 
    DATE_FORMAT(e.encounter_date, '%Y-%m'),
    s.specialty_name
ORDER BY 
    revenue_month,
    total_revenue DESC;

 * mysql+pymysql://root:***@localhost:3306/
1 rows affected.


[("-> Sort: revenue_month, total_revenue DESC  (actual time=51.6..51.6 rows=168 loops=1)\n    -> Stream results  (actual time=47.6..51.5 rows=168 loops= ... (1680 characters truncated) ...      -> Single-row index lookup on s using PRIMARY (specialty_id=p.specialty_id)  (cost=0.25 rows=1) (actual time=914e-6..934e-6 rows=1 loops=7000)\n",)]

**Schema Analysis:**

- **Tables joined**: 4 (billing → encounters → providers → specialties)
- **Number of joins**: 3

**Performance:**

- **Execution time**: 0.0516 seconds (51.6 milliseconds)
- **Estimated rows scanned**: 10,000 rows from billing table
- **Actual rows processed**: 7,000 rows after filtering

**Bottleneck Identified:**

- **Table scan on billing**: Full table scan of 10,000 rows, then filtered to 7,000 rows
- **Multiple sorts**: Data sorted twice (once before grouping, once after aggregation)
- **JOIN chain overhead**: Each of the 7,000 billing records requires 3 index lookups
- **Date calculation at query time**: `DATE_FORMAT()` computed 7,000 times during GROUP BY
- **No indexes on claim_status**: Filtering happens after full table scan

---

## Part 3: Star Schema Design

Design your dimensional model here.

In [ ]:
# Your star schema DDL will go here

---

## Part 4: Performance Comparison

Compare OLTP vs Star Schema query performance.

In [ ]:
# Performance comparison analysis

---

## Cleanup